packages/apollo-server-core/src/requestPipeline.ts

```diff
@@ -162,35 +164,64 @@ export async function processGraphQLRequest<TContext>(
     requestContext,
   });
 
-  const parsingDidEnd = await dispatcher.invokeDidStartHook(
-    'parsingDidStart',
-    requestContext,
-  );
-
   try {
-    let document: DocumentNode;
-    try {
-      document = parse(query);
-      parsingDidEnd();
-    } catch (syntaxError) {
-      parsingDidEnd(syntaxError);
-      return sendErrorResponse(syntaxError, SyntaxError);
+    // If we're configured with a document store, we'll utilize the hash of the
+    // operation in order to lookup a previously parsed-and-validated operation
+    // from that.  A failure to retrieve anything from the cache simply means
+    // we're guaranteed to do the parsing and validation ourselves.
+    let document: DocumentNode | undefined;
+
+    if (config.documentStore) {
+      document = await config.documentStore.get(queryHash);
     }
 
-    requestContext.document = document;
+    // If we still don't have a document, we'll need to parse and validate it.
+    // With success, we'll attempt to save it into the store for future use.
+    if (!document) {
+      const parsingDidEnd = await dispatcher.invokeDidStartHook(
+        'parsingDidStart',
+        requestContext,
+      );
 
-    const validationDidEnd = await dispatcher.invokeDidStartHook(
-      'validationDidStart',
-      requestContext as WithRequired<typeof requestContext, 'document'>,
-    );
+      try {
+        requestContext.document = document = parse(query);
```

Comment:
This line puts the parsed document on requestContext, but the "retried from store" version doesn't. Is that intentional?

Reply:
Good catch! Fixed via f33ae19f6d54aae554432349a99768b3980e5b63.